Bump Traceur to 0.0.82

jmm opened the pull request:
This is a work in progress, but I wanted to get the discussion going. It appears to now be necessary to pass a `sourceRoot` param to traceur to get correct paths in source maps. So I think it's time to make the options to `es6ify.configure()` more general or turn this into a plugin. In this pull request I've implemented the former approach, turning the `filePattern` param into `opts` (which can currently take `filePattern` and `sourceRoot` properties). It falls back to the previous behavior if a regex is passed. (I duck typed it based on `ignoreCase` -- cool?) I rearranged some of the internal options passing based on that change and the addition of `sourceRoot` to the mix.

jmm committed:
Bump traceur to 0.0.82.

jmm committed:
Pass Float32Array to vm re: joyent/node#9121.

jmm committed:
Rearrange options passing.

* Generalize es6ify() argument from filePattern => opts, with
  fallback to previous behavior if a regex is passed. Now filePattern
  should be specified as {filePattern: ...}.

* Support opts.sourceRoot param to pass to traceur. It now appears
  to be neccessary to specifify this to traceur to get correct paths
  in source map.

jmm commented:
I forgot to mention, it may be desirable to default `sourceRoot` to browserify's `basedir`, which would be facilitated by making this a plugin, although I don't think that's advertised as part of the API (which I think it ought to be, for plugins to utilize.)

thlorenz commented:
Nice work @jmm!
Does  a6aafc6 break backwards compat since you rearranged some arg order there?

WRT plugin, would totally love a PR that makes this work as plugin while keeping it working as transform as well. Something similar was [done(among other fixes) to proxyquireify a while ago](https://github.com/thlorenz/proxyquireify/pull/9/files).

jmm commented:
Thanks @thlorenz!

> Does `a6aafc6` break backwards compat since you rearranged some arg order there?

It shouldn't. As long as `compile.js`'s `compileFile()` is only called from `index.js` it should be all good. If it does, then some new tests must be needed to catch it. Is there any expectation that `compile.js` should be usable on its own outside of `index.js` depending on it?

> WRT plugin, would totally love a PR that makes this work as plugin while keeping it working as transform as well. Something similar was done(among other fixes) to proxyquireify a while ago.

I might be able to work on that. Thanks for the link, I'll take a look. I've done some work along those lines (turning a transform into a plugin) on other projects. How do you feel about reading browserify's `basedir` if it's not a documented part of the API though?

jmm committed:
Pass traceur appropriate out file name for map.

jmm committed:
Rough cut at test updated for sourceRoot.

WIP.

jmm commented:
I pushed a couple more commits that are oriented toward getting the correct file name in the source map (e.g. not exposing absolute local paths). They are hasty and rough and definitely need more review. There'll probably need to be separate tests for with and without `sourceRoot` specified.

domenic commented on index.js:
```diff
@@ -23,16 +23,23 @@ function getHash(data) {
-function es6ify(filePattern) {
-  filePattern =  filePattern || /\.js$/;
+function es6ify(opts) {
+  opts || (opts = {});
```
Please use a proper `if` or `?:`. Also best to follow ES6 and do `if (opts === undefined) { opts = {}; }` instead of using any falsy value.

thlorenz replied:
How about just `opts = opts || {}`? 
Unless I"m missing something could `opts` ever be anything but an object?

domenic replied:
Not ES6ey enough, only `undefined` should trigger defaults :)

thlorenz replied:
OK, let's agree to disagree on this one then. I feel the verbosity isn't worth the es6iness.
However doesn't matter enough to me to fight you on that :)

In either case we should make this consistent across the repo, no matter which way we pick.

jmm replied:
> Also best to follow ES6 and do if (opts === undefined) { opts = {}; } instead of using any falsy value.
> Not ES6ey enough, only undefined should trigger defaults :)

I don't see what it accomplishes -- the new test will still allow a bunch of nonsensical values, and convert `es6ify(null)` (which admittedly doesn't make sense) to a `TypeError` -- but it's up to you guys. Also, if making this codebase ES6ey is an imperative, why not actually author it in ES6 and transpile it? 

> Please use a proper `if` or `?:`

You're referring to readability preferences not functionality, right? A `?:` in this situation seems like the worst option to me.

domenic replied:
Yes, TypeError sounds great, and yes, readability.

jmm replied:
Ok.

domenic commented on index.js:
```diff
@@ -23,16 +23,23 @@ function getHash(data) {
+
+  // This function's argument used to be filePattern (regex), so duck type opts
+  // to see if it's meant as filePattern.
+  if (typeof opts.ignoreCase !== 'undefined') opts = {filePattern: opts};
```
I'd say let's just do a major version bump and don't bother with backward compat here.

jmm replied:
I'll proceed on that premise unless I hear otherwise. Since that will make it more verbose to specify `filePattern`, do you want to take this opportunity to shorten that (`patt` maybe)?

domenic replied:
No, filePattern is fine.

jmm replied:
Ok.

domenic commented on index.js:
```diff
@@ -23,16 +23,23 @@ function getHash(data) {
+  // to see if it's meant as filePattern.
+  if (typeof opts.ignoreCase !== 'undefined') opts = {filePattern: opts};
+
+  var filePattern = opts.filePattern = opts.filePattern || /\.js$/;
```
Again use `=== undefined` instead of a falsy test.

jmm replied:
Ok. That will of course allow non-sensical falsy values through.

domenic replied:
Sure, but that's better than disallowing `"0"` or `"false"`.

jmm replied:
You mean `0` or `false`, right? How so? -- those aren't appropriate values for this variable. `(0).test()` or `false.test()` aren't going to work out very well.

domenic replied:
Ideally we'd be converting to a regex or throwing on non-regexes.

jmm replied:
In that case I'd vote for throw.

domenic replied:
Go for it. Allow only `undefined` or a regex.

jmm replied:
Will do.

jmm replied:
Do you have a preference for how to detect this, e.g. duck type it on something like `ignoreCase` or test the value of `Object.prototype.toString.call()`?

domenic replied:
`instanceof RegExp` works.

jmm replied:
:facepalm: Yes, of course.

domenic commented on test/transform.js:
```diff
@@ -23,7 +23,17 @@ test('transform adds sourcemap comment and uses cache on second time', function
+    var file = path.join(sourceRoot, relPath);
+    if (! opts.sourceRoot) relPath = path.basename(file);
+
+    var contents = fs.readFileSync(file).toString();
```
Better to pass `{ encoding: 'utf-8' }` as the second option, that read the whole thing as a buffer then convert it to string with `toString`.

jmm replied:
Good point, I'll fix that.

domenic commented on test/transform.js:
```diff
@@ -23,7 +23,17 @@ test('transform adds sourcemap comment and uses cache on second time', function
+    es6ify = es6ify.configure(opts);
+
+    var file = path.join(sourceRoot, relPath);
+    if (! opts.sourceRoot) relPath = path.basename(file);
```
Style:

``` js
if (opts.sourceRoot === undefined) {
  relPath = path.basename(file);
}
```

jmm replied:
Where you commented on the `opts` arg I assume you're saying you want it to behave like this ES6 code:

``` js
function es6ify (opts = {})
```

But setting aside for the moment any debate over that, what's the point of going to this level of verbosity within functions?

domenic replied:
I don't understand your verbosity argument. This is just better style.

jmm replied:
`'if (! opts.sourceRoot)'.length === 22`
`'if (opts.sourceRoot === undefined)'.length === 34`

Unlike the `filePattern` case, this is one where there may be an argument for allowing `0` or `''` as the value of `sourceRoot`, but I disagree that it's better style when the distinction between `undefined` and falsy is irrelevant. If that's what you want that's what I'll do though.

domenic replied:
Oh, character-counting is how you measure verbosity, OK. And yes, falsy values should not trigger defaults.

jmm replied:
> And yes, falsy values should not trigger defaults.

Why not? (This is a serious question.)

domenic replied:
Because `0` and `false` and `null` and `""` are all meaningful values in many, if not all, cases. And having a uniform rule (`undefined` always triggers defaults) is better than picking on a case-by-case basis.

jmm replied:
I completely disagree, but will bear in mind your preferences when contributing here.

domenic commented on test/transform.js:
```diff
@@ -23,7 +23,17 @@ test('transform adds sourcemap comment and uses cache on second time', function
+    var sourceRoot = path.join(__dirname, '..', 'example', 'src');
+    var relPath = path.join('features', 'iterators.js');
+
+    var opts = {sourceRoot: sourceRoot};
```
Style: spaces after `{` and before `}`

jmm replied:
Thanks for pointing that out. Like I said I worked on those changes hastily, so I could push them before the weekend. I'll fix it.

domenic commented:
Left some comments. Overall looks good, although I am curious about taking this opportunity to convert to a plugin. (I don't see a need for it to be a transform if it is a plugin, personally.) In that case we should also automatically include the runtime, as it's basically nonsensical to use Traceur without it.

It sounds like the source map filename stuff is a mess so I'll trust you to be doing the right thing there. Wish Traceur just did the right thing automatically. Maybe they should try copying 6to5.

jmm commented:
@domenic Thanks for the feedback. I'll fix anything that needs fixing.

> Overall looks good, although I am curious about taking this opportunity to convert to a plugin. (I don't see a need for it to be a transform if it is a plugin, personally.)

I think you're saying you're in favor and you'd prefer it to be only a plugin from now on (breaking backward compatibility by removing the option to use it as a transform)? The only reason I'd give for making it a plugin currently is to access browserify's `basedir` as a default for `sourceRoot`, though I must caution that it's not a documented part of the API. If I'm not mistaken both you and @thlorenz are maintainers on browserify though, right?

> In that case we should also automatically include the runtime, as it's basically nonsensical to use Traceur without it.

I'm currently using it on a project purely to transpile ES6 module syntax, for which I don't need or want the runtime.

> It sounds like the source map filename stuff is a mess so I'll trust you to be doing the right thing there. Wish Traceur just did the right thing automatically. Maybe they should try copying 6to5.

Well I think the addition of the `sourceRoot` param is probably an improvement. I'll be looking into the filename stuff more and report back here on what I find out.

jmm committed:
Make style changes requested in #80.

jmm committed:
Eliminate backcompat for filePattern arg.

jmm committed:
Throw on defined filePattern that's not RegExp.

jmm committed:
Store paths in paths.in / .out.

jmm committed:
Factor out repetitive code.

jmm committed:
Extract to function for use in multiple tests.

jmm committed:
Utilize useSourceRoot.

jmm committed:
Factor out repetitive path massaging code.

jmm committed:
Eliminate hideous string concatenation.

jmm committed:
Add readFileCache().

jmm committed:
Use addsSourceMap() in tests.

jmm committed:
Update documentation.

Individual properties still need descriptions.

jmm commented:
[`.on('error', console.error)`](https://github.com/thlorenz/es6ify/blob/627e999ff6c2aaeb73f150687513ddb4dfb905e7/test/transform.js#L37) or [`.on('error', function (e) { throw e; })`](https://github.com/thlorenz/es6ify/blob/627e999ff6c2aaeb73f150687513ddb4dfb905e7/test/transform.js#L91)?

jmm committed:
Make file reading async.

domenic commented:
For a test? The latter seems better.

jmm committed:
Change error handler.

Per
https://github.com/thlorenz/es6ify/pull/80#issuecomment-72733207

jmm committed:
Refactor to if-else instead of ternaries.

jmm commented:
> For a test? The latter seems better.

Yes, both snippets are in the test file I linked to.

jmm commented:
Here's the rundown on how the paths work out:
- es6ify 1.6.0 / traceur 0.0.79:
  Absolute local file system paths are exposed in `sources`. (es6ify / traceur  expose the absolute local path in `file` also, but that ends up being overwritten by browserify).
- es6ify / traceur 0.0.82
  By default (and at best) fairly useless basenames are populated in `sources`. If there are multiple files (in different directories, of course) with the same basename, then they'll overwrite each other (e.g. if you have `src/a.js` that requires `src/sub/a.js` that requires `src/sub/sub/a.js` you'll wind up with `sources: ["a.js"]` and `sourcesContent` for one of the files. (Ditto about es6ify / traceur exposing the local path in `file` but browserify overwriting.)
- es6ify after these changes are done / traceur 0.0.82
  I have to get more feedback about how traceur itself works before this can be properly resolved. There's a new `sourceRoot` argument to the compiler in the mix, but I'm not sure it's being applied properly. See google/traceur-compiler/issues/1676. Presumably some combination of `sourceName` and `sourceRoot` arguments to traceur will eventually allow generating the desired source map.

In all cases the resulting source map output by browserify has `file: "generated.js", sourceRoot: ""`.

domenic commented:
Thanks so much for tackling this source file mess.

I did some research and [6to5ify makes you pass a base path to it](https://github.com/6to5/6to5ify#relative-source-maps). Maybe we should copy their UX and approach? At this point I kind of just want to bypass Traceur's filename handling as much as possible. WDYT?

jmm commented:
Thanks.

I think that what I have put together right now (which I can probably push later today) is going to be workable despite issues with what traceur is doing, but I hope that traceur will straighten out those issues and then es6ify can be adjusted accordingly. I think getting sane behavior will require passing a base path, either as the traceur `sourceRoot` param (since browserify will overwrite it in the output and it's not needed because of use of `sourcesContent`) or to be used by es6ify to generate a relative path for traceur's `sourceName` param. I tried to get clarification in google/traceur-compiler#1676 about what sort of value `sourceName` is intended to be, but so far no response. It's currently possible to influence traceur's `sources` output by tailoring what you pass as `sourceName` and / or `sourceRoot`.

jmm committed:
Adjust to take basedir opt instead of sourceRoot.

Traceur misappropriates sourceRoot to apply to local input paths and
that shouldn't be propagated to consumers of this package. See
https://github.com/google/traceur-compiler/issues/1676

jmm commented:
See the latest commits for where I'm at now. Tests pass after rebaselining them to current traceur output (although I refactored `test/transform.js` more extensively). These are the highlights:

Instead of `@traceur/generated/TemplateParser/1` in `sources`, there's `@traceur/generated/TemplateParser/2` and `@traceur/generated/TemplateParser/3`. The path for the input file is also changed to match updated traceur output.

So there's `sourcesContent` of `void 0` for `@traceur/generated/TemplateParser/3`. I don't know what's up with that -- if it hints at some problem or not. Frankly I'm not eager to figure it out, but let me know if you think it's important or if the tests don't pass for you.

I changed `es6ify()` to accept an opts hash instead of a `filePattern` arg and originally had a `sourceRoot` property to correspond to traceur's `sourceRoot` argument to it's compile function. But then I realized traceur is misappropriating `sourceRoot` to apply to local input paths and I don't want to propagate that here, so I changed the property to `basedir`. I'm currently passing `basedir` as traceur's `sourceRoot` arg and using it to calculate a relative path for the `sourceName` arg. Note that with es6ify 1.6.0 / traceur 0.0.79 just doing that for `sourceName` could result in relative paths in `sources`. See https://github.com/jmm/es6ify/compare/rel-source-name?expand=1. I don't know exactly what form `sourceName` is intended to take though or if passing an absolute vs relative URL makes any other difference than the `sources` output. See google/traceur-compiler/issues/1676 for discussion of all of that, but I wouldn't hold your breath for answers.

I'm not sure exactly how to format the jsdoc for the opts hash. Is `@param {{filePattern: (undefined|!RegExp), basedir: (undefined|string)}=} opts` the right format here, and if so how do you provide descriptions for the properties? Is docme supposed to be a dev dependency for es6ify?

jmm commented:
I forgot to elaborate on `basedir` -- that's the way to get sane paths in the `sources` output. For the file being transformed, the relative path from `basedir` will be generated and populated in `sources`. So say you have `/somedir/src/js/whatever.js` and you do `es6ify.configure({basedir: '/somedir/src'})`, `sources` will be populated with `js/whatever.js`.

I'm currently experiencing a problem in chrome though where those `//# sourceURL=` comments are screwing up display of source files. I haven't had the chance to figure out what's happening there yet.

jmm commented:
Created #81 for further discussion of conversion to a plugin.

domenic commented:
Regarding docme and jsdoc, that's all @thlorenz, I am not a jsdoc fan and just try to copy what I see around me :P.

The `basedir` scheme sounds good, although I wonder if a name involving source maps might be more transparent. 6to5ify uses "sourceMapRelative".

jmm commented:
> Regarding docme and jsdoc, that's all @thlorenz, I am not a jsdoc fan and just try to copy what I see around me :P.

Ok, thanks. Hopefully @thlorenz can point me in the right direction.

> The `basedir` scheme sounds good, although I wonder if a name involving source maps might be more transparent. 6to5ify uses "sourceMapRelative".

Like I said, until the last minute I had it as `sourceRoot`, but I no longer believe that makes sense. Then I thought of `basedir`, because browserify's `basedir` might be a sensible default (haven't thought that all the way through yet). I think another name could be ok, but I don't find `sourceMapRelative` very descriptive (does that come from anywhere else, a spec or anything?). I think a name more indicative of source maps but more descriptive would be `sourcesBase` or something like that.

domenic commented:
sourceMapBase or sourceMapBaseDir?

jmm commented:
I sort of feel like prefixing any source map-related opts with `sourceMap` is going to do more harm than good unless the names are made really verbose, like `sourceMapSourcesBaseDir`. I think that's what the documentation is for, to explain that an option like `sourcesRoot` relates to the source map. If it's really necessary to prefix them, perhaps `sm` would be sufficient as a hint? E.g. `smBaseDir`? I think if everything would need to be prefixed with `sourceMap` I'd be tempted to make `sourceMap` a hash that can contain properties like `sourcesRoot`, e.g. `b.plugin(es6ify, {sourceMap: {sourcesRoot: '...'}})`

thlorenz commented:
Possibly @guzart wants to join in WRT to API choices as he was working on a major API update himself in #25.
We should also revisit that PR to get some API ideas from there.

thlorenz commented:
@jmm added you as collab, so once we agree on everything you could squash and _forward_ merge into master.
Also now you can close issues (I suppose some questions will come up related to the API changes once they are in). Hoping for you to triage some of those.
Thanks for the good work so far.

jmm commented:
I'm going to leave some notes here for future reference.

The current pull request brings traceur up to 0.0.82 with proper `sources` paths in the output (when `basedir` [or whatever it ends up being called] is supplied), but the way that's achieved is hackneyed because of traceur's misapplication of its `sourceRoot` input param to alter the values that end up in `sources`. It looks like traceur has [come around](https://github.com/google/traceur-compiler/issues/1676#issuecomment-72995017) to stop doing that. Once that happens the solution should probably be to pass the exact desired `sources` output to traceur's compile function (the `sourceName` arg). (Which might've been possible for es6ify to do before traceur added the misapplied `sourceRoot` param.) So probably the values passed as `sourceName` and `outputName` will be generated with something like `path.resolve(opts.basedir, file)`. It may or may not be necessary to pass a `sourceRoot` argument to traceur (perhaps with value `false`). Unless someone can think of a use case for populating the source map generated by traceur with a `sourceRoot` value, I think the ideal thing is to have it omitted (either by passing nothing or `false` to traceur depending how that shakes out), because:
- `sourcesContent` is populated.
- Browserify overwrites it anyway.

So if that all works out there probably shouldn't be separate tests anymore for with / without the `sourceRoot` argument to traceur, but maybe there should be tests for with / without a `basedir` argument to es6ify.

jmm commented:
> Possibly @guzart wants to join in WRT to API choices as he was working on a major API update himself in #25.
> We should also revisit that PR to get some API ideas from there./

Ok. Sorry @guzart, didn't realize you'd already done so much work in this direction. I skimmed #25 (but haven't looked at any of the commits) and it sounds similar to what I've been working on in #81. I think I'll clean up what I have so far a little but and push it so everyone can take a look and if @guzart wants to take the reins again can cherry pick anything good from there. One thought I do have right off the bat is that I think `addTraceurRuntime` is unnecessarily verbose. I currently have `includeRuntime`, which is barely better (maybe `addRuntime` is better than both), but I don't think there'll be any confusion about which runtime it is.

jmm commented:
> @jmm added you as collab, so once we agree on everything you could squash and forward merge into master.

Ok, thanks. I kind of think the history is useful, but it's up to you. If you want me to squash, all into one commit?

> Also now you can close issues (I suppose some questions will come up related to the API changes once they are in). Hoping for you to triage some of those.

Ok, I'll see what I can do.

> Thanks for the good work so far.

Thanks.

thlorenz commented:
@jmm squash as you see fit.
A lot of commits here were good to drive the discussion, but I believe there should be up to 5 things that were added/changed.
So squashing to <=5 commits seems sensible.

domenic commented:
>  I think addTraceurRuntime is unnecessarily verbose. I currently have includeRuntime, which is barely better (maybe addRuntime is better than both), but I don't think there'll be any confusion about which runtime it is.

Since runtimes will be on by default it should be false by default, so e.g. `omitRuntime: true` or `noRuntime: true` would be the way to get rid of the runtime.

jmm commented:
I think a double negative is more confusing (do not do it? yes) than a straightforward: do it? -- yes or no -- that defaults to yes. But, it's up to you.

domenic commented:
Yeah I feel pretty strongly that boolean options should have a default value of false. If you can think of a better phrasing that's cool (e.g. maybe "omit" is better than "no").

jmm commented:
I think no matter what you name it it's more confusing than the straightforward version. I'll let you know if I have a suggestion for a default: `false` name.

jmm commented:
> Possibly @guzart wants to join in WRT to API choices as he was working on a major API update himself in #25.
> We should also revisit that PR to get some API ideas from there.

I pushed more commits to the branch I linked in #81 to show what I have so far.

jmm commented:
FYI, traceur 0.0.83 is out now and partially fixes the problems with `sourceRoot` and `sources`. See also google/traceur-compiler/issues/1695 google/traceur-compiler/issues/1696

domenic commented:
So there's been a lot of back-and-forth. At some point we should merge and do a release :). What are the unresolved issues remaining? (Apologizes if you're waiting on me or @thlorenz to respond to something that got lost in the shuffle...)

guzart commented:
unfortunately :( there isn't much I can bring in to the conversation, main points have already been discussed:
- include traceur runtime by default
- forward traceur options via `traceurOverrides`
- and simply allow to pass options directly via the main method instead of using `es6ify.configure`

jmm commented:
Ok, I think these are the unresolved issues:

For this pull request:
- What to call the param currently known as `basedir`. See discussion in this thread. I favor something like `basedir`, `sourcesBase`, or `smBasedir` (or `smBaseDir`). If source map-related options need to be segregated, I could also see passing a `sourceMap` hash containing options like `basedir` or `sourcesBase`.
- Finer points of the JSdoc comments and regeneration of the docs. Maybe @thlorenz can point me in the right direction.
- Make sure all relevant documentation is updated.
- Squashing the commits.

Here's a summary of issues that have been discussed in or are related to this thread that relate to conversion to a plugin and should be hashed out in #25 or #81:
- Whether or not to default the `basedir` param (whatever it ends up being called) to browserify's `basedir`.
- What to call the option for controlling inclusion of the runtime.
- How to include the runtime.
- Whether to eliminate `compileFile` export.

jmm commented:
@guzart Do you want to jump back in and make the actual commits that are going to get merged?

There's also the question of how to include the runtime. It looks like you were taking a different approach. I've posted more about that in #81.

domenic commented:
@jmm 

> What to call the param currently known as basedir

Can you explain why `sourcesBase` makes sense to you? What does "sources" even mean?

My reasoning for something more explicit like "sourceMapBaseDir" is that (a) I don't mind if people type extra characters; (b) it's very clear what it does; (c) it explicitly lets you know this is only related to source maps and not to all "sources" or to es6ify in general.

jmm commented:
> Can you explain why sourcesBase makes sense to you? What does "sources" even mean?

It means the `sources` array in the source map. The parameter is to be used to do this to generate the value that will be populated in `sources`:

``` js
path.relative(basedir, file)
```

Where `file` is the absolute path browserify passes to es6ify (or any transform).

> My reasoning for something more explicit like "sourceMapBaseDir" is that (a) I don't mind if people type extra characters; (b) it's very clear what it does; (c) it explicitly lets you know this is only related to source maps and not to all "sources" or to es6ify in general.

A) In my opinion, all else being equal (including descriptiveness), shorter is preferable.

B) I'm not so sure it is. It's not used in any other context (like the source map spec or mozilla/source-map), is it?, and there are a number of paths / URLs related to source maps: `sources`, `sourceRoot`, `file`, `sourceMappingURL`, `sourceURL` (, others?) that may or may not be influenced by this param. I wouldn't count on users knowing what it does without reading the documentation. I actually do want to apply it to the `file` property of the source map as well, although that'll be overwritten by browserify in normal usage. I'm looking into what's up with the `//# sourceURL=` parts -- it should probably be applied to those as well.

C) I guess. There's only a handful of parameters for es6ify and the user isn't going to know what they are or how to use them in the first place without looking at the documentation.

adbl commented:
We are experience these issues, is this PR going in or are you waiting for further traceur updates? including traceur 0.0.83 which includes other related changes?

@jmm is there any workaround for the time being to get a working (in chrome) source map?

jmm commented:
@adbl 

> We are experience these issues, is this PR going in or are you waiting for further traceur updates? including traceur 0.0.83 which includes other related changes?

I personally am waiting for some combination of traceur maturing its API with respect to source maps...
https://github.com/google/traceur-compiler/issues/1695
https://github.com/google/traceur-compiler/issues/1715
https://github.com/google/traceur-compiler/issues/1748
...

and continuation of discussion here about the unresolved issues in this thread.

> is there any workaround for the time being to get a working (in chrome) source map?

Here's the workaround I'm using to deal with https://github.com/google/traceur-compiler/issues/1748 / [chromium # 459682](https://code.google.com/p/chromium/issues/detail?id=459682) (which was bogusly merged with another issue):

``` js
var rs = require('readable-stream');
var stream = new rs.Transform;

stream.data = '';

stream._transform = function (chunk, enc, cb) {
  this.data += chunk;
  cb();
};

stream._flush = function (cb) {
  this.data = this.data.replace(/\/\/# sourceURL=\S+/mg, '');
  this.push(this.data);
  cb();
};

return b
  .transform(es6ify)
  .bundle()
  .pipe(stream)
```

And here's the workaround I'm using to correct the paths in the source map:

``` js
gulp.task('build:fix-source-maps', ['build:bundle'], function (done) {
  if (! use_source_map) return;
  var map_file = './dist/rsrc/app.js.map';

  fs.readFile(map_file, function (err, src) {
    if (err) throw err;
    var map = JSON.parse(src);
    map.sources = map.sources.map(function (source) {
      var
        base_dir = [__dirname, 'src'],
        ret = source;

      [
        base_dir,
        base_dir.slice(0, -1),
        base_dir.slice(-1)
      ].every(function (base_dir) {
        base_dir = path.join.apply(path, base_dir);
        if (ret.indexOf(base_dir + '/') === 0) {
          ret = path.relative(base_dir, ret);
          return false;
        }
        return true;
      });

      return ret;
    });

    fs.writeFile(map_file, JSON.stringify(map, null, 2), done);
  });
});
```
